lib/at_cmd/at_cmd.c

```diff
@@ -160,7 +162,10 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 
 			LOG_ERR("Unrecoverable reception error (err: %d), "
 				"thread killed", errno);
+
+			atomic_clear_bit(&running, 0);
 			close(common_socket_fd);
+			k_mem_slab_free(&rsp_work_items, (void **)&item);
 			return;
```

Comment:
I need to understand, this can only happen in case this thread failed to recover the socket on its own (by simply closing and reopening it). Does it really help if we do the same manually with a call to `at_cmd_driver_init`? How can the application tell that this thread is down and it needs to re-initialize it with `at_cmd_driver_init`?

Reply:
The recover attempt will fail because during the shutdown process the interfaces (the different socket types) in bsdlib is brought which cocecently leads to the open() call to fail. So to avoid having a mem slab in allocated state I need to free it here so the whole pool is in free state. 
Actually I think the close on line 167 could be removed, it will anyway be forced shut by the shutdown process.

Reply:
You cannot restart a thread :)
You'll need to run at beginning of the thread function when you expect to be able to open a socket again.

Reply:
Just wondering how should we handle this problem. Assuming that some operation will fail when we disable bsdlib, so we can recover later seems error-prone to me.

I think we should provide `*_deinit` function for all modules that depend on bsdlib to operate (`at_cmd` and all modules that rely on it). This way, an application could do a graceful shutdown of all of the dependent modules, before shutting down bsdlib.

And as @lemrey says, we should not try to respawn a thread. While this might work, it's not a functionality supported by the kernel, and thus can lead to some undefined behaviour. I've finally found some docs to back me up here:
https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-termination
i.e. 
> The kernel does not currently make any claims regarding an application’s ability to respawn a thread that terminates.

Reply:
> Just wondering how should we handle this problem. Assuming that some operation will fail 
> when we disable bsdlib, so we can recover later seems error-prone to me.

I inject that error into the socket on shutdown to signal that the link is down and then I wake all sleeping threads to make them process this, so it's not an assumption. 

But ok, I see, I was assuming that the kernel was freeing the threads context and that it could be respawned after it had died. Seems like I can use the k_thread_join() function and join the thread with a shutdown thread in the os integration code that can exit on the next init. This way I don't need to exit this thread and can make it continue on my next init. The pid to that thread could be passed through a deinit function.

I can make a try and update the PR later today.